internal_ws/yktrace/src/swt.rs

```diff
@@ -86,23 +61,10 @@ struct SwtLoc {
     bb_idx: u32
 }
 
-const TL_TRACE_INIT_CAP: usize = 1024;
-const TL_TRACE_REALLOC_CAP: usize = 1024;
-
-/// The trace buffer.
-///
-/// `calloc` and `free` are directly used instead of `Vec` to avoid calling into traced functions
-/// from `__yk_swt_rec_loc`. Otherwise there would be infinite recursion.
-#[thread_local]
-static mut TRACE_BUF: *mut SwtLoc = 0 as *mut SwtLoc;
-
-/// The number of elements in the trace buffer.
-#[thread_local]
-static mut TRACE_BUF_LEN: usize = 0;
-
-/// The allocation capacity of the trace buffer (in elements).
-#[thread_local]
-static mut TRACE_BUF_CAP: usize = 0;
+thread_local! {
+    /// The trace buffer.
+    static TRACE_BUF: RefCell<Vec<SwtLoc>> = RefCell::new(Vec::with_capacity(1024));
```

Comment:
Is it trivial to prove that two borrows cannot occur? If so, you might be able to be faster without runtime borrow checks? Maybe I'm jumping the gun a bit...

Reply:
It shouldn't be possible for two borrows to occur. I could use `UnsafeCell` if you prefer.

Reply:
Since this is quite performance critical, I think we could allow it, but I'd like to check with @ltratt.

Laurie: For context, this vector is the buffer used to collect a software trace. There will be one push per recorded block. How do you feel about `UnsafeCell`?

Reply:
Is the `RefCell` here just to make things mutable within a thread local?

Reply:
Yes

Reply:
OK. I agree that the frequency with which we call this might make `RefCell` a noticeable overhead. My suggestion would be that we create a new type that wraps the `Vec` and does the `UnsafeCell` trickery itself, rather than having lots of `unsafe` stuff sprinkled around everywhere.

Reply:
Done in a76df1ee80d27bbb405905b357df4283dacc3b50.